proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,302 @@
+validator repairs all blocks which could potentially have been optimistically
+confirmed.
+
+3. After repair is complete, the validator counts votes on each fork and
+sends out local heaviest fork.
+
+4. Each validator counts if enough nodes can agree on one block (same slot and
+hash) to restart from:
+
+  1. If yes, proceed and restart
+
+  2. If no, print out what it thinks is wrong, halt and wait for human
+
+See each step explained in details below.
+
+### 1. Gossip last vote before the restart and ancestors on that fork
+
+The main goal of this step is to propagate the locally selected fork to all
+others in restart.
+
+We use a new Gossip message `LastVotedForkSlots`, its fields are:
+
+- `last_voted_slot`: `u64` the slot last voted, this also serves as last_slot
+for the bit vector.
+- `last_voted_hash`: `Hash` the bank hash of the slot last voted slot.
+- `ancestors`: `BitVec<u8>` compressed bit vector representing the slots on
+sender's last voted fork. the most significant bit is always
+`last_voted_slot`, least significant bit is `last_voted_slot-81000`.
+
+The number of ancestor slots sent is hard coded at 81000, because that's
+400ms * 81000 = 9 hours, we assume most restart decisions to be made in 9 
+hours. If a validator restarts after 9 hours past the outage, it cannot join 
+the restart this way. If enough validators failed to restart within 9 hours, 
+then fallback to the manual, interactive cluster restart method.
+
+When a validator enters restart, it uses ephemeral shred version to avoid
+interfering with those outside the restart. There is slight chance that 
+the ephemeral shred version would collide with the shred version after the
+silent repair phase, but even if this rare case occurred, we plan to flush the 
+CRDS table on successful restart, so gossip messages used in restart will be 
+removed.
+
+### 2. Repair ledgers up to the restart slot
+
+The main goal of this step is to repair all blocks which could potentially be
+optimistically confirmed.
+
+We need to prevent false negative at all costs, because we can't rollback an 
+optimistically confirmed block. However, false positive is okay. Because when 
+we select the heaviest fork in the next step, we should see all the potential 
+candidates for optimistically confirmed slots, there we can count the votes and
+remove some false positive cases.
+
+However, it's also overkill to repair every block presented by others. When
+`LastVotedForkSlots` messages are being received and aggregated, a validator
+can categorize blocks missing locally into 3 categories: ignored, must-have, 
+and unsure. Depending on the stakes of validators currently in restart, some
+slots with too few stake can be safely ignored, some have enough stake they
+should definitely be repaired, and the rest would be undecided pending more
+confirmations.
+
+Assume `RESTART_STAKE_THRESHOLD` is 80% and that 5% restarted validators can
+make mistakes in voting.
```

Comment:
"can make mistakes in voting." - being more precise about the exact voting concern here I think would be helpful for overall understanding. I think we're trying to point out the risk of validators literally changing their votes (deliberately or otherwise) in the `LastVotedForkSlots` message from what they actually voted on pre-restart

Reply:
Changed.